File: crates/chthl/src/generator.rs
//! This module defines the HTML Generator, which walks the AST produced by the
//! parser and generates an HTML string.

use crate::ast::{CommentNode, ElementNode, Expression, Node, Program, TextNode};

pub fn generate_program(program: &Program) -> String {
    let mut html = String::new();
    for node in &program.nodes {
        html.push_str(&generate_node(node, 0));
    }
    html
}

// The `indent` parameter is for pretty-printing, which can be useful for debugging.
fn generate_node(node: &Node, indent: usize) -> String {
    match node {
        Node::Element(el) => generate_element_node(el, indent),
        Node::Text(txt) => generate_text_node(txt, indent),
        Node::Comment(comment) => generate_comment_node(comment, indent),
    }
}

fn generate_element_node(element: &ElementNode, indent: usize) -> String {
    let self_closing_tags: &[&str] = &[
        "area", "base", "br", "col", "embed", "hr", "img", "input", "link", "meta", "param",
        "source", "track", "wbr",
    ];

    let mut tag = format!("{}{}<{}", " ".repeat(indent), "\n", element.tag_name);
    for attr in &element.attributes {
        let value = match &attr.value {
            Expression::Ident(s) => s,
            Expression::StringLiteral(s) => s,
        };
        tag.push_str(&format!(" {}=\"{}\"", attr.name, value));
    }

    if self_closing_tags.contains(&element.tag_name.as_str()) {
        tag.push('>');
        return tag;
    }
    tag.push('>');

    let mut children_html = String::new();
    for child in &element.children {
        children_html.push_str(&generate_node(child, indent + 2));
    }

    if children_html.is_empty() {
         format!("{}</{}>", tag, element.tag_name)
    } else {
        format!("{}{}{}</{}>", tag, children_html, "\n", element.tag_name)
    }
}

fn generate_text_node(text: &TextNode, indent: usize) -> String {
    // In a real implementation, we would handle HTML escaping here.
    format!("{}{}{}", " ".repeat(indent), "\n", text.value)
}

fn generate_comment_node(comment: &CommentNode, indent: usize) -> String {
    format!("{}{}\n<!-- {} -->", " ".repeat(indent), "\n", comment.value)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;
    use crate::parser::Parser;

    #[test]
    fn test_html_generation() {
        let input = r#"
            html {
                head {
                    title { text { "My CHTL Page" } }
                }
                body {
                    id: "app";
                    -- main content
                    div {
                        class: "container";
                        h1 { text { "Welcome!" } }
                        img { src: "/logo.png"; }
                    }
                }
            }
        "#;

        let lexer = Lexer::new(input);
        let mut parser = Parser::new(lexer);
        let program = parser.parse_program();
        assert!(parser.errors().is_empty());

        let html = generate_program(&program);

        // Note: The expected output is based on the pretty-printing logic (indents and newlines)
        // in the generator. This makes the test specific but easier to read.
        let expected_html = r#"
<html>
  <head>
    <title>
      My CHTL Page
    </title>
  </head>
  <body id="app">
    <!-- main content -->
    <div class="container">
      <h1>
        Welcome!
      </h1>
      <img src="/logo.png">
    </div>
  </body>
</html>"#;

        // A simple normalization to make comparison less brittle regarding whitespace.
        let normalize = |s: &str| s.lines().map(|l| l.trim()).filter(|l| !l.is_empty()).collect::<String>();

        assert_eq!(normalize(html.as_str()), normalize(expected_html));
    }
}

File: chtl_compiler/generator.py
from .ast_nodes import ProgramNode, ElementNode, TextNode, AttributeNode, StyleBlockNode

class HtmlGenerator:
    def generate(self, node: ProgramNode) -> str:
        if not isinstance(node, ProgramNode):
            return self._visit(node)

        return "".join(self._visit(child) for child in node.children)

    def _visit(self, node):
        if node is None:
            return ""
        method_name = f'_visit_{node.__class__.__name__}'
        visitor = getattr(self, method_name, self._generic_visit)
        return visitor(node)

    def _generic_visit(self, node):
        raise Exception(f'No _visit_{node.__class__.__name__} method for {node}')

    def _visit_ElementNode(self, node: ElementNode) -> str:
        attributes = " ".join(self._visit(attr) for attr in node.attributes)

        style_str = ""
        if node.style_block:
            props = [f"{prop.name}: {prop.value};" for prop in node.style_block.properties]
            style_str = f' style="{" ".join(props)}"'

        open_tag_parts = [node.tag_name]
        if attributes:
            open_tag_parts.append(attributes)

        open_tag = f"<{' '.join(open_tag_parts)}{style_str}>"

        self_closing_tags = {'area', 'base', 'br', 'col', 'embed', 'hr', 'img', 'input', 'link', 'meta', 'param', 'source', 'track', 'wbr'}
        if node.tag_name.lower() in self_closing_tags:
            return open_tag

        body = "".join(self._visit(child) for child in node.children)

        close_tag = f"</{node.tag_name}>"

        return f"{open_tag}{body}{close_tag}"

    def _visit_AttributeNode(self, node: AttributeNode) -> str:
        return f'{node.name}="{node.value}"'

    def _visit_TextNode(self, node: TextNode) -> str:
        return node.content

    def _visit_StyleBlockNode(self, node: StyleBlockNode) -> str:
        # Style blocks are handled by the ElementNode visitor, so they don't produce output themselves.
        return ""
